dttools/src/list.c

```diff
@@ -1,406 +1,633 @@
 /*
-Copyright (C) 2003-2004 Douglas Thain and the University of Wisconsin
-Copyright (C) 2005- The University of Notre Dame
+Copyright (C) 2018- The University of Notre Dame
 This software is distributed under the GNU General Public License.
 See the file COPYING for details.
 */
 
+#include <assert.h>
+#include <limits.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
 #include "list.h"
 
-#include <stdlib.h>
+struct list {
+	unsigned refcount;
+	unsigned length;
+	struct list_item *head;
+	struct list_item *tail;
+	struct list_cursor *iter;
+};
+
+struct list_item {
+	unsigned refcount;
+	double priority;
+	struct list *list;
+	struct list_item *next;
+	struct list_item *prev;
+	void *data;
+};
 
-struct list *list_create(void)
-{
-	struct list *l;
+struct list_cursor {
+	struct list *list;
+	struct list_item *target;
+};
 
-	l = malloc(sizeof(struct list));
-	if(!l)
-		return 0;
+static void oom(void) {
+	const char *message = "out of memory\n";
+	write(STDERR_FILENO, message, strlen(message));
+	abort();
+}
+
+static void list_ref(struct list *list) {
+	if (!list) return;
+	assert(list->refcount < UINT_MAX);
+	++list->refcount;
+}
 
-	l->head = 0;
-	l->tail = 0;
-	l->size = 0;
-	l->iter = 0;
+static void list_unref(struct list *list) {
+	if (!list) return;
+	assert(list->refcount > 0);
+	--list->refcount;
+}
 
-	return l;
+static void list_item_ref(struct list_item *item) {
+	if (!item) return;
+	assert(item->refcount < UINT_MAX);
+	++item->refcount;
 }
 
-struct list *list_splice(struct list *top, struct list *bottom)
-{
-	if(!top->head) {
-		list_delete(top);
-		return bottom;
+static void list_item_unref(struct list_item *item) {
+	if (!item) return;
+	assert(item->refcount > 0);
+	if (--item->refcount == 0) {
+		if (item->prev != item)
+			list_item_unref(item->prev);
+		list_item_unref(item->next);
+		free(item);
 	}
+}
+
+static void list_cursor_relax(struct list_cursor *cur) {
```

Comment:
What's this function about?

Reply:
This comes back to the per-item refcounts. `list_drop()` flags items as dead by tying their `->prev` pointer back to themselves and removes a reference. Rather than keeping a list of pointers to update, this function gets called whenever someone touches a pointer. It checks if the item is dead and fixes things up. Changing to 1 writer xor n readers would eliminate all this.